### PR TITLE
Add speculative decoding example to OpenVINO quantized generation notebook

### PR DESCRIPTION
**Pull request by ofirzaf**
# What does this PR do?

Add support and example for running speculative decoding generation with OpenVINO.

This PR is depending on PR #620 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: notebooks/openvino/quantized_generation_demo.ipynb [hunk 1]
```diff
@@ -223,14 +223,342 @@
...
+   "id": "5179c53d-0436-4ee9-9367-2625a8d3e262",
+   "metadata": {},
+   "source": [
+    "## Assisted generation\n",
+    "Auto-regressive language models generate outputs token by token. Assisted generation (AG) is a general name for a group of methods that speculate the next generated tokens and then use the language model to validate the speculated tokens and accept/reject them.\n",
+    "AG is a great method to accelerate LMs running locally on your computer as it reduces memory bandwidth requirements and can speedup generation by 1.5x-3x without any accuracy degradation.\n",
+    "You can read more on assisted generation here in this great [blog post](https://huggingface.co/blog/assisted-generation).\n",
+    "\n",
+    "\n",
+    "In this section we will present how to run Phi-2 with 2 AG methods that are well supported withing 🤗 transformers, Prompt Lookahead Decoding (PLD) and Speculative Decoding.\n",
```
Comment (echarlaix) [inline notebooks/openvino/quantized_generation_demo.ipynb:237]:
```suggestion
    "In this section we will present how to run Phi-2 with two AG methods that are well supported within 🤗 transformers : Prompt Lookahead Decoding (PLD) and Speculative Decoding.\n",
```

File: notebooks/openvino/quantized_generation_demo.ipynb [hunk 2]
```diff
@@ -223,14 +223,342 @@
...
+   "cell_type": "markdown",
+   "id": "98d34b03-55e0-4606-be26-5722d6868679",
+   "metadata": {},
+   "source": [
+    "### Prompt lookahead decoding\n",
+    "Now we will run the same example from before with PLD enabled. \n",
+    "PLD speculates tokens by searching the last n-gram (usually 2-gram) in the sequence inside the prompt, if we find a match, we will take the next few tokens (configured with `prompt_lookup_num_tokens`) as our speculation, if a match is not found the code will revert back to auto-regressive generation.\n",
+    "\n",
+    "We will run the same example from before with PLD. To enable PLD, we simply pass the `prompt_lookup_num_tokens` key-word argument to the `generate` function.\n",
+    "Note that PLD can be great when doing code completion as some sequences of tokens tend to repeat themselves in the same order, names of variables, `for i in range(...):`, etc.\n"
```
Comment (echarlaix) [inline notebooks/openvino/quantized_generation_demo.ipynb:283]:
```suggestion
    "Note that PLD can be great when doing code completion as some sequences of tokens tend to repeat themselves in the same order, names of variables, like `for i in range(...):`, etc.\n"
```

File: notebooks/openvino/quantized_generation_demo.ipynb [hunk 3]
```diff
@@ -223,14 +223,342 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1466938c-0945-4eb6-a80f-dd165cc5eca1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "asst_model.generation_config.num_assistant_tokens = 3\n",
+    "asst_model.generation_config.num_assistant_tokens_schedule = 'const'"
```
Comment (echarlaix) [inline notebooks/openvino/quantized_generation_demo.ipynb:398]: would prefer double quotes to stay consistent

File: notebooks/openvino/quantized_generation_demo.ipynb [hunk 4]
```diff
@@ -223,14 +223,342 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dab6669b-f3f1-411e-b4b8-31ead823247f",
+   "metadata": {},
+   "source": [
+    "Note that in both cases of AG we presented, the generation result is exactly the same as Phi-2 would have generated it without AG!\n",
+    "\n",
+    "Like we mentioned before, the AR is a very important metric for choosing a draft.\n",
```
Comment (echarlaix) [inline notebooks/openvino/quantized_generation_demo.ipynb:432]: to improve clarity we can add :
```suggestion
    "Like we mentioned before, the acceptance rate (AR) is a very important metric for choosing a draft.\n",
```

File: notebooks/openvino/quantized_generation_demo.ipynb [hunk 5]
```diff
@@ -223,14 +223,342 @@
...
+    "        return np.mean(ar_per_win)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c35f5e0c-5ed6-4011-a295-80a81fea8b8e",
+   "metadata": {},
+   "source": [
+    "Now we can use any dataset for text generation task and measure the AR on that dataset.\n",
+    "Here we use the HumanEval dataset for evaluating code generation.\n",
```
Comment (echarlaix) [inline notebooks/openvino/quantized_generation_demo.ipynb:522]:
```suggestion
    "Here we use the [HumanEval](https://huggingface.co/datasets/openai_humaneval) dataset for evaluating code generation.\n",
```

File: notebooks/openvino/quantized_generation_demo.ipynb [hunk 6]
```diff
@@ -223,14 +223,342 @@
...
+    "We run the model with speculative decoding on 30 samples and calculate the AR.\n",
```
Comment (echarlaix) [inline notebooks/openvino/quantized_generation_demo.ipynb:523]:
```suggestion
    "We run the model with speculative decoding on 30 samples.\n",
```

**Review by echarlaix — APPROVED** [review]
Looks good, thanks for the addition @ofirzaf !

**Comment by ofirzaf** [comment]
Hi @echarlaix, updated the notebook per your suggestions. Thanks!

**Review by echarlaix — APPROVED** [review]

**Pull request merged**